pandas/core/reshape.py

```diff
@@ -558,87 +549,138 @@ def _convert_level_number(level_num, columns):
+    new_levels += level_vals
+    new_labels += [np.tile(labels, N) for labels in zip(*level_labels)]
+    new_names += level_vals_used.names
+    new_index = MultiIndex(levels=new_levels, labels=new_labels,
+                           names=new_names, verify_integrity=False)
+
+    # if stacking all levels in columns, result will be a Series
+    if len(frame.columns.levels) == num_levels_to_stack:
+        new_data = frame.values.ravel()
+        if dropna:
+            mask = notnull(new_data)
+            new_data = new_data[mask]
+            new_index = new_index[mask]
+        return Series(new_data, index=new_index)
+
+    # result will be a DataFrame
+
+    def _equal_key(keya, keyb):
+        return all(((a == b) or
+                    ((isinstance(a, float) and np.isnan(a) and isinstance(b, float) and np.isnan(b)) or
+                     (isinstance(a, NaTType) and isinstance(b, NaTType))))
+                   for a, b in zip(keya, keyb))
+
+    # construct new_columns
+    if len(frame.columns.levels) > (num_levels_to_stack + 1):
+        # result columns will be a MultiIndex
+        # tuple list excluding level for grouping columns
         tuples = list(zip(*[
-            lev.take(lab) for lev, lab in
-            zip(this.columns.levels[:-1], this.columns.labels[:-1])
+            _make_new_index(lev, lab) for lev, lab in
+            zip(this.columns.levels[:-num_levels_to_stack],
+                this.columns.labels[:-num_levels_to_stack])
         ]))
-        unique_groups = [key for key, _ in itertools.groupby(tuples)]
-        new_names = this.columns.names[:-1]
+        unique_groups = [x for i, x in enumerate(tuples) if (i == 0) or not _equal_key(x, tuples[i - 1])]
+        new_names = this.columns.names[:-num_levels_to_stack]
         new_columns = MultiIndex.from_tuples(unique_groups, names=new_names)
     else:
-        new_columns = unique_groups = this.columns.levels[0]
+        # result columns will be an Index
+        new_columns = _make_new_index(this.columns.levels[0], sorted(set(this.columns.labels[0])))
+        unique_groups = list(new_columns)
 
-    # time to ravel the values
+    # construct new_data
     new_data = {}
-    level_vals = this.columns.levels[-1]
-    level_labels = sorted(set(this.columns.labels[-1]))
-    level_vals_used = level_vals[level_labels]
-    levsize = len(level_labels)
     drop_cols = []
-    for key in unique_groups:
-        loc = this.columns.get_loc(key)
-        slice_len = loc.stop - loc.start
+    try:
+        nan_index = level_vals_used.get_loc(np.nan)
+        if not isinstance(nan_index, int):
+            nan_index = None
+    except KeyError:
+        nan_index = None
+    unique_group_levels = this.columns.nlevels - num_levels_to_stack
+    columns_values = this.columns.get_values()
+    for i, key in enumerate(unique_groups):
+        # The following nonsense is needed to properly handle Nans. Otherwise would simply do:
```

Comment:
Again trying to retain/regain my sanity. The original code consists of the two commented out lines. Because ``.get_loc(key)`` doesn't properly handle a ``key`` that is a tuple containing ``NaN``s, I had to come up with this beautiful code.